server/stream.go

```diff
@@ -919,7 +919,7 @@ func (mset *Stream) processInboundJetStreamMsg(_ *subscription, pc *client, subj
 	numConsumers := len(mset.consumers)
 	interestRetention := mset.config.Retention == InterestPolicy
 
-	var resp = &JSPubAckResponse{}
+	var resp = &JSPubAckResponse{PubAck: &PubAck{Stream: name}}
```

Comment:
I wonder if we should only attach this PubAck inside the code where detect an error? Avoid creating and throwing away in normal case.

Reply:
Yeah, can do, did think this is a bit of a waste here, but wanted to avoid accidentally forgetting it on one.  Will change round

Reply:
instead of allocating `resp` each time and then discarding it in the happy path:

```go
resp.PubAck = &PubAck{Stream: name}
resp.Error = &ApiError{Code: 400, Description: err.Error()}
b, _ = json.Marshal(resp)
sendq <- &jsPubMsg{reply, _EMPTY_, _EMPTY_, nil, b, nil, 0}
```

I was thinking maybe this?

```go
errorResp = func(c int, d string) []byte {
  resp := &JSPubAckResponse{
    PubAck: &PubAck{Stream: name},
    Error: &ApiError{Code: c, Description: d}
  }

  b, _ := json.Marshal(resp)
  return b
}

.
.
.

sendq <- &jsPubMsg{reply, _EMPTY_, _EMPTY_, nil, errorResp(400, "foo"), nil, 0}
```

Reply:
Would not make it a closure inside the processInboundStream message but if external func for errors yes.

Reply:
The cost of setting up the closure function may be worse then the original allocation strategy ;)

Reply:
Hehe could be not sure how clever the compiler is :) anyway will just set this in every call and hope we don’t make a mistake

Reply:
Sounds good.